modules/class-simplechart-save.php

```diff
@@ -53,10 +48,71 @@ function save_post_action( $post_id ) {
 
 		remove_action( 'save_post_simplechart', array( $this, 'save_post_action' ), 10, 1 );
 		$post = get_post( $post_id );
-		$this->do_save_post( $post );
+		$this->_do_save_post( $post );
 		add_action( 'save_post_simplechart', array( $this, 'save_post_action' ), 10, 1 );
 	}
 
+	protected function _do_save_post( $post ) {
+		// verify nonce
+		if ( empty( $_POST['simplechart-nonce'] ) || ! wp_verify_nonce( sanitize_text_field( wp_unslash( $_POST['simplechart-nonce'] ) ), 'simplechart_save' ) ) {
+			return;
+		}
+
+		// delete featured image if post is NOT published but has a featured image
+		if ( 'publish' !== get_post_status( $post->ID ) && has_post_thumbnail( $post->ID ) ) {
+			wp_delete_attachment( get_post_thumbnail_id( $post->ID ), true );
+		}
+
+		// handle base64 image string if provided
+		if ( ! empty( $_POST['save-previewImg'] ) ) {
+			$this->_save_chart_image( $post, sanitize_text_field( wp_unslash( $_POST['save-previewImg'] ) ), $this->_default_img_type );
+		}
+
+		// handle raw CSV multiline data
+		if ( ! empty( $_POST['save-rawData'] ) ) {
+			add_filter( 'sanitize_text_field', array( $this, 'sanitize_raw_data' ), 99, 2 );
+			$this->_save_chart_image( $post, sanitize_text_field( wp_unslash( $_POST['save-rawData'] ) ), $this->_default_img_type );
+			remove_filter( 'sanitize_text_field', array( $this, 'sanitize_raw_data' ), 99, 2 );
+		}
+
+		// handle JSON fields
+		foreach ( array( 'chartData', 'chartOptions', 'chartMetadata' ) as $field ) {
+			if ( ! empty( $_POST[ 'save-' . $field ] ) ) {
+				// sanitize field name w/ esc_attr() instead of sanitize_key() because we want to preserve uppercase letters
+				update_post_meta( $post->ID, 'save-' . esc_attr( $field ), sanitize_text_field( wp_unslash( $_POST[ 'save-' . $field ] ) ) );
+			}
+		}
+
+		// save error messages
+		if ( ! empty( $this->_errors ) ) {
+			update_post_meta( $post->ID, 'simplechart-errors', $this->_errors );
+		}
+
+		// save debug messages
+		if ( ! empty( $this->_debug_messages ) ) {
+			update_post_meta( $post->ID, 'simplechart-debug', $this->_debug_messages, true );
+		}
+	}
+
+	/**
+	 * Allow sanitize_text_field() for multiline CSV without stripping "\n" chars
+	 *
+	 * @param string $filtered Filtered input
+	 * @param string $initial Unfiltered input
+	 * @return string Sanitized multiline csv or empty string
+	 */
+	public function sanitize_raw_data( $filtered, $initial ) {
+		// strip newlines from initial input
+		$initial_stripped = preg_replace( '/\n+/', ' ', $initial );
+		if ( $filtered === $initial_stripped ) {
+			// if that's the only difference, then the initial string is fine
+			return $initial;
+		} else {
+			// if anything else was removed by sanitize_text_field(), return an empty string
+			return '';
```

Comment:
might you want an error here? i could see someone (i.e. any reporter) trying to include an HTML tag in their data and being 😩  and 😠 when it fails